pkg/tarfs/tarfs.go

```diff
@@ -362,33 +362,85 @@ func (t *Manager) blobProcess(ctx context.Context, snapshotID, ref string,
 			err = t.generateBootstrap(ds, snapshotID, layerBlobID, upperDirPath)
 			if err != nil && !errdefs.IsAlreadyExists(err) {
 				return epilog(err, "generate tarfs data from image layer blob")
-			} else {
-				msg := fmt.Sprintf("Nydus tarfs for snapshot %s is ready", snapshotID)
-				return epilog(nil, msg)
 			}
+			msg := fmt.Sprintf("Nydus tarfs for snapshot %s is ready", snapshotID)
+			return epilog(nil, msg)
 		}
 	}
 
 	keyChain, err := auth.GetKeyChainByRef(ref, nil)
 	if err != nil {
-		epilog(err, "create key chain for connection")
-		return err
+		return epilog(err, "create key chain for connection")
 	}
 	remote := remote.New(keyChain, t.insecure)
 	rc, _, err := t.getBlobStream(ctx, remote, ref, layerDigest)
 	if err != nil && remote.RetryWithPlainHTTP(ref, err) {
 		rc, _, err = t.getBlobStream(ctx, remote, ref, layerDigest)
 	}
 	if err != nil {
-		epilog(err, "get blob stream for layer")
-		return errors.Wrapf(err, "get blob stream by digest")
+		return epilog(err, "get blob stream for layer")
 	}
 
-	go process(rc, remote)
+	if retry {
+		// Download and convert layer content in synchronous mode when retry for error recovering
+		err = process(rc, remote)
+	} else {
+		// Download and convert layer content in background.
+		// Will retry when the content is actually needed if the background process failed.
+		go func() {
+			_ = process(rc, remote)
+		}()
+	}
 
 	return err
 }
 
+func (t *Manager) retryPrepareLayer(snapshotID, upperDirPath string, labels map[string]string) error {
+	ref, ok := labels[label.CRIImageRef]
+	if !ok {
+		return errors.Errorf("not found image reference label")
+	}
+	layerDigest := digest.Digest(labels[label.CRILayerDigest])
+	if layerDigest.Validate() != nil {
+		return errors.Errorf("not found layer digest label")
```

Comment:
Maybe wrap the error of `Validate()`?  The `err != nil` does not indicate  `not found layer digest label`.

Reply:
Fixed